### ISSUE TITLE
[return-await] False negative for arrow functions in Promise.all

### ISSUE THREAD

**Issue by bcuccioli** [typescript-eslint/typescript-eslint#3331]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/return-await": ["error", "always"]
  }
}
```

```TS
async function f(n: number) {
  return n+1;
}

async function g() {
  const l = [1,2,3];
  return await Promise.all(l.map(n => f(n)));
}
```

**Expected Result**

I expected the inner function `n => f(n)` to be caught by the lint rule; it should be `async n => await f(n)` under this rule.

**Actual Result**

No lint error.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.22.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `15.14.0` |

**Comment by alonstern** [comment]
IIUC, the bug here is that, in general, [return-await] does not support non-async functions. e.g.:
```
async function f(n: number) {
  return n+1;
}

function g() {
  return f(2)
}
```
would not raise a lint error.

Is anybody currently working on this? Can I take this one?

**Comment by bradzacher** [comment]
It seems I misread and misunderstood the original issue.

The issue here is that you have created a non-async function that returns a promise.
This case is specifically **not** handled by the rule.

The rule *only* checks promise returns within `async` functions.

If you want a rule to enforce that your functions that return promises are marked as `async`, then you can use [`promise-function-async`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/promise-function-async.md).
